### ISSUE TITLE
USER ISSUE: Cert verify failed: UNITYTLS_X509VERIFY_FLAG_NOT_TRUSTED

### ISSUE THREAD

**Issue by Elliander** [StrangeLoopGames/EcoIssues#15417]
**Version:** 0.8.3.3 beta

**Steps to Reproduce:**
Try to login

**Expected behavior:**
Should login, as older versions do no problem

**Actual behavior:**
Doesn't login.

**Do you have mods installed? Does issue happen when no mods are installed?:**
No mods.

This is similar to this issue:
https://github.com/StrangeLoopGames/EcoIssues/issues/14579

Except that the certificate that is supposed to be deleted there does not exist on my computer at all. Additionally, I am still able to login to v0.7.3.3.3 so the assumption that this cannot be fixed on your end is false, since it's a bug introduced at some point since that version. Even if this is something that the user can fix on their end, it's still a very bad thing to happen as the average non-tech savvy user won't really know what to do here.

I'll check every version going back until I figure out where it starts and report back. In the meantime, here's a screenshot of the error:

<a href="https://ibb.co/8PbXNDt"><img src="https://i.ibb.co/8PbXNDt/failed-to-login.jpg" alt="failed-to-login" border="0"></a>

and this is the text copied to clipboard when I click "copy text and log"

> Failed to login
> 
> Unknown Error
> 
> Log File:
> Initialize engine version: 2019.2.17f1 (8e603399ca02)
> [XR] Discovering subsystems at path C:/EcoPC_v0.8.3.3-beta/Eco_Data/UnitySubsystems
> GfxDevice: creating device client; threaded=1
> Direct3D:
>     Version:  Direct3D 11.0 [level 11.1]
>     Renderer: NVIDIA GeForce GTX 980 Ti (ID=0x17c8)
>     Vendor:   
>     VRAM:     6097 MB
>     Driver:   26.21.14.3064
> <RI> Initializing input.
> 
> <RI> Input initialized.
> 
> <RI> Initialized touch support.
> 
> UnloadTime: 0.381673 ms
> Curl error 51: Cert verify failed: UNITYTLS_X509VERIFY_FLAG_NOT_TRUSTED
> 
>  
> (Filename: C:\buildslave\unity\build\Modules/UnityWebRequest/Implementations/TransportCurl.cpp Line: 799)
> 
> Curl error 51: Cert verify failed: UNITYTLS_X509VERIFY_FLAG_NOT_TRUSTED
> 
>  
> (Filename: C:\buildslave\unity\build\Modules/UnityWebRequest/Implementations/TransportCurl.cpp Line: 799)
> 
>

**Comment by Elliander** [comment]
I checked version version going back (all 64-bit), and the last version of Eco that works without this error is: EcoPC_v.0.8.2.9-beta

When I launch that version, it doesn't even ask me to login. It somehow remembers my login from previous versions launched from my computer and displays past servers played on. So 8.3 asks me to login, despite having already logged in on this machine and having login info saved, while 8.2 doesn't give me the option of logging in it just does so automatically and works. 

I unfortunately cannot test the login prompt for this version on my computer as a result, but I wonder if the problem is at all related to having already logged in successfully in a prior version or not.

**Comment by SlayksWood** [comment]
1. Do you have Firewall or Antivirus enabled?
2. Please provide your Operating System

**Comment by Elliander** [comment]
1.) No Firewall or antivirus running. 
2.) Windows 10 64-bit

I'm updating my server to v0.8.2.9 since that's the last known good version, but I'll test any changes you make.

**Comment by Elliander** [comment]
Also: I noticed that in these later versions, a seed number is removed as an option from WorldGenerator.eco. I added it back in, but I am not sure if it does anything now. It also regenerated an already generated world. Not sure why it would do that, since it means servers that already exist can't really be upgraded.

**Comment by Elliander** [comment]
I read up on this issue, and found an answer here:

https://stackoverflow.com/questions/59945581/unity-tlsexception-handshake-failed-unitytls-x509verify-flag-not-trusted

Where the issue was caused by a broken chain of trust preventing the verification of an otherwise valid certificate. The solution being to provide the entire chain.

I've read quite a few examples of this error over the last day, where in each one the developers did not encounter it on their end, but end users started to. 

As stated by unity, this is the result of a deliberate decision to allow cross platform support:

https://issuetracker.unity3d.com/issues/error-of-tlsexception-when-system-dot-net-dot-http-dot-dll-is-used-for-http-in-the-editor

My guess here is that either Unity changed something between version 8.2.9 and 8.3 (are you using a different version of Unity?) or you changed something that broke the chain. Possibly both?

I read over here that a change to Unity coinciding with a change by Microsoft resulted in this error on Hololens products:

https://forum.unity.com/threads/uniy-2018-2-https-webrequest-failes-on-latest-hololens-os-udpate.550429/

So it seems to be a fairly common problem type which prevents end users from logging into software, which is very bad. You definitely don't want there to be even a chance of this error cropping up in a final release.

One thought I have though is why it's affecting me and not necessarily others? Could it be the case that wherever my prior login information is stored is contributing to the problem? I can test that out by installing on a friend's computer who never played Eco before, but if I do I'd want to know how to delete the login credentials from that computer. Once I have that information I can carry out that test, and if it works I can see about removing (preferably, after backing up) the login credentials on my main computer and then testing it out to see if it continues to be a problem. (On a related note, why isn't there an option to logout? I imagine that on shared computers like the classrooms you market to that would be very important to have.)

I'd do more to help, but I haven't done much work in the Unity engine myself yet.

**Comment by mirasrael** [comment]
Hi. Thanks for detailed info, actually we already had this problem before, but unfortunately nothing we can fix on our side. But you can follow the instruction to fix this on your configuration:

https://github.com/StrangeLoopGames/EcoIssues/wiki/UNITYTLS_X509VERIFY_FLAG_NOT_TRUSTED-issue-troubleshooting

**Comment by Elliander** [comment]
I'll take a look at those instructions, but I'm not convinced that this is something that can't be fixed on your end. The last thing you want is software that may or may not run on a given client machine, especially if you hope to push this to schools as the teachers of classrooms will have much less control over their machines.

**Comment by Elliander** [comment]
**Short Response:** 

This hasn't worked for me.

**Long Response:**

The instructions could be more clear for those users who may not be as computer literate. Under Chrome it says to check the certificates in Chrome, but it doesn't say how to. 

To do this, you go to the web page and then you open the menu (the 3 dots) and click "More Tools" then "Developer Tools". Then click on the ">>" on the far right of the new menu bar until you see "Security". Click that and then under "Overview" click "View Certificate". Then click "Certificate Path" to show you the image that this guide otherwise shows.

The average user would not know how to do that.

and yes, I had "Let's Encrypt Authority X1" so I moved onto the next set of instructions.

"reset cert cache (in elevated prompt)" is another instruction that the average user won't understand.

To actually do this (on Windows 10) you type cmd.exe in the search box at the bottom of the start menu, but don't press enter. Instead look at where it appears in the menu above and right-click "run as administrator". This will open the command prompt with elevated permissions.

Unfortunately, using both of these commands in order did not fix the problem for me. Trying to play the game again resulted in the same error with the same error text. 

I went back to play.eco , refreshed the page, and clicked to view the certificate again and it was still on Let's Encrypt Authority X1, so I figured I needed to restart CertSCV so typed:

net stop CertSCV && net start CertSCV

Into command prompt, but it couldn't find that service. Then I checked services.msc and it wasn't listed there at all. My next step will be to try a reboot after I'm done with some work I'm doing, but beyond that I'm not sure what else I can try.

Once I know all the steps required, I could probably write a simple C++ application that checks for a specific certificate, clears it if found, launches a webpage (if that's an important step), and whatever else is required in one batch operation to fix this easily for others without having to go through these steps.

**Comment by Elliander** [comment]
Nothing seems to be working. The certificate is still X1.

The first command is supposed to set the certificates to expire, but it just gets the same certificate again and it's set to expire in 2021. 

The second command is supposed to delete it, but I get this error when I try:

```
CertUtil: -URLCache command FAILED: 0x80070103 (WIN32/HTTP: 259 ERROR_NO_MORE_ITEMS)
CertUtil: No more data is available.
```

Rebooting doesn't make a difference either. Repeating the commands through several reboots makes no difference either.

What's maddening here is that the certificate doesn't appear to be stored where Windows 10 keeps it's other certificates, so I can't manually delete it by typing "certificates" into run, and it's also not showing up in the the list of certificates under settings in either Chrome or Firefox so I can't delete it there.

There has to be some way to manually delete these. 

I can still login to 8.2.9, so why don't you guys go back to using the old certificate system until you can find a permanent fix for this problem? Any bug that prevents end users from playing at all regardless of what they do is literally a game breaking bug.

**Comment by redwyre** [comment]
You may need to check both user certificates ("Manage user certificates") and machine certificates ("Manage computer certificates").

Both the Let's Encrypt authority and Unity's security policy are out of our control. Fortunately it's rare that anyone gets hit with this issue.

**Comment by Elliander** [comment]
I wish I read your response half an hour ago, because that's the conclusion I finally reached myself.

In any case, I finally freaking got it fixed!

OK, so the certificate is stored not under Console Root, but under "Current User".

Type MMC.exe into run and press enter.
File > add/remove snap in > Certificates
Click add
Select "My user account"
Click Finish

Then you will find "Let's encrypt authority X1" under "Intermedia Certification Authorities > Certificates

Right-click and delete it.

However, this wasn't enough. I then had to go to:

"Trusted Root Certification Authorities" > "Certificates" > 

and right-click delete "DST Root CA X3" in both Console Root AND current User. Once all 3 certificates were manually deleted play.eco was able to issue the correct X3 certificate. Then I was able to play in 0.8.3.3 at long last.

Since deleting 3 certificates is the solution, I imagine the simplest approach would be to have eco delete these certificates if this error is encountered and then try to login again, which it could do seamlessly in the background. 

Some background as to why this is an issue: X3 uses the same identifier as X1, so the browsers don't know to get a new one. Whereas they should have revoked the certificate when moving to X3 they were lazy. That makes IdenTrust as much to blame for this mess as Unity.
